### ISSUE TITLE
Track Channel Profit

### ISSUE THREAD

**Issue by carlaKC** [lightningnetwork/lnd#3645]
Routing nodes that need to allocate scarce capital will want to know how much profit each of their channel is making so that they can efficiently manage capital, have visibility on the amount their capital is earning and potentially close channels that are not profitable. 

Dependent on #3420, this can be achieved by tracking the HTLCs that are forwarded through a channel. HTLC forwards arrive on an incoming channel and depart on an outgoing channel, so profit from a forward must be attributed to both channels to prevent underestimating the influence of mostly-inbound channels. 

Two related measures that operators may be interested in:
1. Utility: The portion of funds moving along a channel which are for sends from the node or receives to it. This gives a measure of how much personal use the node gets from a channel, despite those HTLCs not earning fees.
2. Failed forwards: The portion of HTLCs which are forwarded though the channel and then later failed. This is not an actionable metric, but provides interesting insight into how often capital is being locked up (and how long it is locked up for) without earning a fee.

**Comment by halseth** [comment]
To get historical data, can we use the open channel's revocation log to extract the information about which HTLCs got settled/failed and the profitability? The information should be there and would allow us to extract this information for all existing channels

**Comment by carlaKC** [comment]
I've been having a look at this today and I think that querying the forwarding log, as in `rpcserver.go#ForwardingHistory`, would be a good place to start? Since we want to attribute profit across channels, it would be nice to have the information by channel pair? Our receives won't be included in this query, but that information is available elsewhere. 

Since I'd expect we'll be doing these calculations for channels that are open for a long time, I'm wondering whether to query periodically to keep running track of profit in memory (we can discard the events once processed and query DB just with an offset from then) or to just process on demand. 

**Rough Calcs**
Transaction volumes per day (as per LNBig's twitter reports):
Regular: 200 
Upper bound: 3500
and for completeness, a low tx count of 5 a day

Size of DB query for regular/upper tx volumes by length of query period, given a forwarding event size of [40 bytes](https://github.com/lightningnetwork/lnd/blob/master/rpcserver.go#L4728):

1 month:
- Low; 05 * 31 * 40 bytes =.0062 MB
- Regular: 200 * 31 * 40 bytes = 0.248 MB
- Upper: 3500 * 31 * 40 bytes = 4.34 MB

6 months:
- Low: 5 * 182 * 40 bytes = 0.0364
- Regular: 200 * 182 * 40 bytes = 1.456 MB
- Upper: 3500 * 182 * 40 bytes = 25.48 MB

1 year:
- Low: 5 * 365 * 40 bytes = 0.073 MB
- Regular: 200 * 365 * 40 bytes =  2.92 MB
- Upper: 3500 * 356 * 40 bytes = 49.84 MB

I don't think that we are going to be seeing anything near 3500 txns a day for a while, but I think it makes sense to periodically update a profit metric in memory rather than calculate in one go on demand?

**Comment by carlaKC** [comment]
Also some thoughts about how we're going to store this profit metric chanfitness side. 

Two approaches:
1. Store stripped down forwarding events in the `chanEventLog` and calculate profit on demand.
pro: users can specify their own time value of BTC % in rpc calls for profit
con: have to store events in memory (timestamp, amount, fee)

2. Store a running profit balance as discard events as we go.
pro:do not have to store events for each channel
con: time value of BTC will be a set % which users can configure on startup, will have to restart LND to change it

**Comment by carlaKC** [comment]
An alternative approach to adding local events would be to rely on existing `listpayments`/`listinvoices` calls. The advantage of this would be that we do not need to extend the forwarder log. 

The disadvantage would come query side, due to the way the payments/invoices buckets are structured:
- Payments: by payment hash, need to list all payments them check whether the first hop in their route is the channel we're calculating profit for
- Invoices: by incremental add index, either list all invoices or do a binary search to find the indexes which bound the time we're interested in then check channel id

**Comment by carlaKC** [comment]
Closing, can be implemented outside of lnd with correct information exposed.
